### PR TITLE
dockerfile: define Envoy image using ARG

### PR DESCRIPTION
**Pull request by EricMountain**
Move the cilium-envoy image reference to an ARG so that it can be overridden via build-args.

This allows overrides using mirrored images for instance and is useful in build environments without Internet access as well as avoiding API throttling on public registries.

```release-note
Defines the cilium-envoy image used in the build Dockerfile using ARG to allow overrides. 
```

### CODE REVIEW

File: images/cilium/Dockerfile
Comment (sayboras) [inline images/cilium/Dockerfile]: One of the reasons that it's not passed as argument is to make sure that the helm chart is in sync with envoy deamonset image. There is also a Makefile target (e.g. make -C image check-envoy-image) to validate the same.

Just curious if you have other use case for custom builds, or just for mirrored image and air-gap environment, also do you need to install Cilium with one extra flag (e.g. disable-envoy-version-check) ?

Comment (jrajahalme) [reply images/cilium/Dockerfile]: There are envoy update scripts that deal with the `FROM` line that need to be adjusted for the `ARG`, I recall Makefiles may do that too.

**Comment by sayboras** [comment]
Thanks for the update, can you help to squash two commits into one? Also, signing off commit is required to make the bot happy.

File: images/cilium/Dockerfile
Comment (EricMountain) [reply images/cilium/Dockerfile]: Thanks for pointing this out, I hadn't realised.

So:
* Updated `{update,check}-cilium-envoy-image.sh` to patch the Dockerfile ARG entry instead of the FROM.
* Updated the link-time flags to pull the Envoy image ref from the ARG entry instead of the FROM.

Hope I haven't missed anything.

Comment (sayboras) [reply images/cilium/Dockerfile]: Yes, I believe the above will be enough to make CI happy.

Comment (EricMountain) [reply images/cilium/Dockerfile]: > Just curious if you have other use case for custom builds, or just for mirrored image and air-gap environment, 

We do custom builds because we typically carry a few patches on top of upstream: mostly these are bug fixes backported from more recent versions or `main`, and occasionnally experiments/dev that hasn't made it upstream yet.

> also do you need to install Cilium with one extra flag (e.g. disable-envoy-version-check) ?

From my understanding that check (`checkEnvoyVersion()`) only verifies the SHA which is unchanged so that option is not needed.

**Comment by EricMountain** [comment]
> Thanks for the update, can you help to squash two commits into one? Also, signing off commit is required to make the bot happy.

@sayboras Got back to this, sorry for the delay. Hopefully this is all good now.

**Review by sayboras — APPROVED** [review]
LGTM :heavy_check_mark:

**Review by sayboras — APPROVED** [review]
LGTM :heavy_check_mark:

**Comment by sayboras** [comment]
Thanks for sharing your use case, as well as the changes.

**Comment by sayboras** [comment]
I have enabled auto merge for this PR i.e. once all approvals are in, it will be merged automatically.

**Comment by EricMountain** [comment]
Rebased due to conflict: a new version of the Envoy image was merged to `main`.

Is there anything I can or should do to obtain the remaining approvals?

**Comment by sayboras** [comment]
> Rebased due to conflict: a new version of the Envoy image was merged

Sorry about this, let me ask around for the missing approval. Thanks for your continuous effort on this change.

**Review by sayboras — APPROVED** [review]

**Review by tklauser — APPROVED** [review]

**Pull request merged**
